src/storage.rs

```diff
@@ -686,6 +692,155 @@ impl ArchetypeData {
+                            let (src, _, _) = other_tags.data_raw();
+                            let dst = self_tags.get_mut(*type_id).unwrap().alloc_ptr();
+                            other_tags.element().clone(src.as_ptr(), dst);
+                        }
+                    }
+                })
+            });
+
+            // Iterate all the chunks within the source chunk set
+            for (_src_chunk_idx, src_chunk) in src_chunk_set.chunks.iter().enumerate() {
+                // Copy the data from source to destination. Continuously find or create chunks as
+                // needed until we've copied all the data
+                let mut entities_remaining = src_chunk.len();
+                while entities_remaining > 0 {
+                    // Get or allocate a chunk.. since we could be transforming to a larger component size, it's possible
+                    // that even a brand-new, empty chunk won't be large enough to hold everything in the chunk we are copying from
+                    let dst_free_chunk_index =
+                        self.get_free_chunk(dst_chunk_set_index, entities_remaining);
+                    let dst_chunk_set = &mut self.chunk_sets[dst_chunk_set_index];
+                    let dst_chunk = &mut dst_chunk_set.chunks[dst_free_chunk_index];
+
+                    // Determine how many entities we will write
+                    let entities_to_write =
+                        std::cmp::min(entities_remaining, dst_chunk.capacity() - dst_chunk.len());
+
+                    // Prepare to write to the chunk storage
+                    let mut writer = dst_chunk.writer();
+                    let (dst_entities, dst_components) = writer.get();
+                    let dst_components = unsafe { &mut *dst_components.get() };
+
+                    // Find the region of memory we will be reading from in the source chunk
+                    let src_entity_start_idx = src_chunk.len() - entities_remaining;
+                    let src_entity_end_idx = src_entity_start_idx + entities_to_write;
+
+                    // Copy all the entities to the destination chunk. The normal case is that we simply allocate
+                    // new entities.
+                    //
+                    // We also allow end-user to specify a HashMap<Entity, Entity>. The key is an Entity from
+                    // the source chunk and the value is an Entity from the destination chunk. Rather than appending
+                    // data to the destination chunk, we will *replace* the data, according to the mapping. This
+                    // is specifically intended for use with hot-reloading data. When some source data is changed,
+                    // we can use the mapping to respawn entities as needed using the new data.
+
+                    // We know how many entities will be appended to this list
+                    dst_entities.reserve(dst_entities.len() + entities_to_write);
```

Comment:
Can this be pre-allocated at the beginning of the loop instead of causing re-growths every iteration?

Reply:
This would require a pre-pass through all chunk sets to gather/allocate all chunks that will be written to. Would you still want to do this?